### PR TITLE
Fix the SQL SELECT query in _paginate_room_events_txn

### LINKED ISSUE #6337: /sync table-scans events
the labels support seems to have introduced a performance regression in the /sync code such that it tries to table-scan events, which is ... suboptimal

### PR DESCRIPTION
**Pull request by babolivier**
Doing a SELECT DISTINCT when paginating is quite expensive, because it requires the engine to do sorting on the entire events table. However, we only need to run it if we're filtering on 2+ labels, so this PR is changing the request so that DISTINCT is only used then.

Fixes #6337

### CODE REVIEW

**Comment by richvdh** [comment]
fixes #6337?

**Comment by babolivier** [comment]
> fixes #6337?

Ah, yes, I missed the issue

File: synapse/storage/data_stores/main/stream.py [hunk 1]
```diff
@@ -871,14 +871,38 @@ def _paginate_room_events_txn(
...
+        select_keywords = "SELECT"
+        join_clause = ""
+        if event_filter and event_filter.labels:
+            # If we're not filtering on a label, then joining on event_labels will
+            # return as many row for a single event as the number of labels it has. To
+            # avoid this, only join if we're filtering on at least one label.
+            join_clause = (
+                "LEFT JOIN event_labels"
+                " USING (event_id, room_id, topological_ordering)"
+            )
```
Comment (erikjohnston) [inline synapse/storage/data_stores/main/stream.py]: Can you please use multiline strings please.

File: synapse/storage/data_stores/main/stream.py [hunk 2]
```diff
@@ -871,14 +871,38 @@ def _paginate_room_events_txn(
...
-            "SELECT DISTINCT event_id, topological_ordering, stream_ordering"
+            "%(select_keywords)s event_id, topological_ordering, stream_ordering"
             " FROM events"
-            " LEFT JOIN event_labels USING (event_id, room_id, topological_ordering)"
+            " %(join_clause)s"
             " WHERE outlier = ? AND room_id = ? AND %(bounds)s"
             " ORDER BY topological_ordering %(order)s,"
             " stream_ordering %(order)s LIMIT ?"
-        ) % {"bounds": bounds, "order": order}
+        ) % {
```
Comment (erikjohnston) [inline synapse/storage/data_stores/main/stream.py]: Can you please use multiline strings please.

**Pull request merged**